fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryManager.java

```diff
@@ -73,6 +87,82 @@ public synchronized void close() throws IOException {
+                props -> props.clearState().setReturnedRowLimit(1));
+
+        final Range range = state.indexSubspace.range();
+        final KeyRange keyRange = new KeyRange(range.begin, range.end);
+        final Subspace subspace = state.indexSubspace;
+        final KeyExpression rootExpression = state.index.getRootExpression();
+
+        if (! (rootExpression instanceof GroupingKeyExpression)) {
+            mergeIndex(analyzerWrapper, TupleHelpers.EMPTY);
+            return AsyncUtil.DONE;
+        }
+        GroupingKeyExpression expression = (GroupingKeyExpression) rootExpression;
+        final int groupingCount = expression.getGroupingCount();
+
+        final RecordCursor<Tuple> cursor = new ChainedCursor<>(
+                state.context,
+                lastKey -> nextTuple(state.context, subspace, keyRange, lastKey, scanProperties, groupingCount),
+                Tuple::pack,
+                Tuple::fromBytes,
+                null,
+                ScanProperties.FORWARD_SCAN);
+
+        return cursor
+                .map(tuple ->
+                        Tuple.fromItems(tuple.getItems().subList(0, groupingCount)))
+                .forEach(groupingKey ->
+                        mergeIndex(analyzerWrapper, groupingKey));
+    }
+
+    private void mergeIndex(LuceneAnalyzerWrapper analyzerWrapper, Tuple groupingKey) {
+        try {
+            getDirectoryWrapper(groupingKey).mergeIndex(analyzerWrapper);
+        } catch (IOException e) {
+            throw new RecordCoreStorageException("Lucene mergeIndex failed", e);
+        }
+    }
+
+    @SuppressWarnings("PMD.CloseResource")
+    private static CompletableFuture<Optional<Tuple>> nextTuple(@Nonnull FDBRecordContext context,
+                                                                @Nonnull Subspace subspace,
+                                                                @Nonnull KeyRange range,
+                                                                @Nonnull Optional<Tuple> lastTuple,
+                                                                @Nonnull ScanProperties scanProperties,
+                                                                int groupingCount) {
+        KeyValueCursor.Builder cursorBuilder =
+                KeyValueCursor.Builder.withSubspace(subspace)
+                        .setContext(context)
+                        .setContinuation(null)
+                        .setScanProperties(scanProperties);
+
+        if (lastTuple.isPresent()) {
+            final byte[] lowKey = subspace.pack(Tuple.fromItems(lastTuple.get().getItems().subList(0, groupingCount)));
+            cursorBuilder
+                    .setLow(lowKey, EndpointType.RANGE_EXCLUSIVE)
+                    .setHigh(range.getHighKey(), range.getHighEndpoint());
+        } else {
+            cursorBuilder.setContext(context)
+                    .setRange(range);
+        }
+
+        return cursorBuilder.build().onNext().thenApply(next -> {
```

Comment:
Since the merge is performed by another transaction (probably in a background queue), the store that performed the user operation is long gone. Even if we'll find a way to preserve this info, it could not be trusted.

Reply:
I agree that this should be a later optimization (if we do it at all), but I think it can be trusted to the point where it can be passed around. This probably isn't the place to discuss how to add more generic support for running delayed work.